[KeyVault] Delete secrets in perf test cleanup

mikeharder opened the pull request.

mikeharder commented on sdk/keyvault/azure-security-keyvault-perf/src/main/java/com/azure/security/keyvault/secrets/perf/core/SecretsTest.java:
```diff
@@ -41,4 +45,13 @@ public SecretsTest(TOptions options) {
+    protected Mono<Void> deleteAndPurgeSecretsAsync(String ... names) {
+        return Flux
+            .fromArray(names)
+            .flatMap(name -> secretAsyncClient
```
The goal here is to ignore exceptions thrown if the secret being deleted does not exist.  Does this reactor code look correct?  It seems to work in a basic test.

vcolin7 replied:
I think it should work. There might be some ways to make this a bit more "reactive" but it should do the work. Any comments @srnagar @alzimmermsft?

srnagar replied:
This looks good to me. The only thing to verify is the semantics of delete and purge methods when the resource is not found. Does it throw `ResourceNotFoundException` or some other type of exception (HttpResponseException?) or return a 200? Also, log the name of the secret that was not found.

```
.beginDeleteSecret(name)
.doOnError(name -> logger.info(name + " not found")
.onErrorResume(ResourceNotFoundException.class, e -> Mono.empty())
```

mikeharder replied:
I don't think `PerfStressTest` has a `logger` object to use, should I just log to `System.out`?

vcolin7 commented on sdk/keyvault/azure-security-keyvault-perf/src/main/java/com/azure/security/keyvault/secrets/perf/ListSecretsTest.java:
```diff
@@ -10,16 +10,43 @@
+        if (secretClient.listPropertiesOfSecrets().iterator().hasNext() ||
+            secretClient.listDeletedSecrets().iterator().hasNext()) {
+
+            throw new RuntimeException("KeyVault " + secretClient.getVaultUrl() + "must contain 0 " +
```
Should we consider deleting all secrets and then continue instead of throwing?

srnagar replied:
Return `Mono.error()` instead.

>Should we consider deleting all secrets and then continue instead of throwing?

@vcolin7 We should not automatically delete just to be safe that we are not deleting something that shouldn't be. The executor of perf tests should ensure that there are no secrets before running the tests. Alternatively, we could add a flag to the test to delete all secrets before starting the test.

mikeharder replied:
@srnagar covered most of it.  It's dangerous for tests to delete resources they didn't create, so we just check and throw instead.  Perf tests should clean up after themselves, so you should only hit this if an earlier perf test crashed (or somehow failed to cleanup), or if you ran against the wrong KeyVault (and we **really** don't want to delete everything then).

srnagar commented on sdk/keyvault/azure-security-keyvault-perf/src/main/java/com/azure/security/keyvault/secrets/perf/ListSecretsTest.java:
```diff
@@ -10,16 +10,43 @@
-                .map(i -> "listSecretsPerfTest-" + UUID.randomUUID())
-                .flatMap(b -> secretAsyncClient.setSecret(b, b))
-                .then());
+        super.globalSetupAsync().block();
```
We should avoid calling `block()` in an async method.

mikeharder replied:
Do you know how to convert the code in this method to reactor?  I used `block()` here because I wasn't sure how to chain all the other code after `globalSetupAsync()`.  One way I would know how to do it, is to move all the code after `globalSetupAsync()` into a separate async method, then chain this after `globalSetupAsync()`.

mikeharder replied:
I attempted to convert it using `then(Mono.defer())`.  Is this the best way?

mikeharder commented:
@vcolin7, @srnagar: I believe I have addressed all feedback.  I also decided to simplify the `deleteAndPurgeSecretsAsync()` method by allowing errors to just propagate, since there should be no errors deleting or purging secrets the test just set.  I originally ported the code from .NET which does ignore errors but I don't think this is correct.

srnagar commented:
>since there should be no errors deleting or purging secrets the test just set

@mikeharder I am not sure how Key Vault library handles retries for delete/purge operations. If a delete operation is retried, we might still get a 404 and the test will fail without cleaning up the rest of the secrets.

mikeharder commented:
> > since there should be no errors deleting or purging secrets the test just set
> 
> @mikeharder I am not sure how Key Vault library handles retries for delete/purge operations. If a delete operation is retried, we might still get a 404 and the test will fail without cleaning up the rest of the secrets.

I don't see any special exception handling in our sample:

https://github.com/Azure/azure-sdk-for-java/blob/main/sdk/keyvault/azure-security-keyvault-secrets/src/samples/java/com/azure/security/keyvault/secrets/ManagingDeletedSecretsAsync.java

Our perf tests should use the same code we recommend customers use.  If customers need to handle 404 retry errors during delete/purge, we should either improve our SDK or samples, and then perf tests can align.

srnagar commented:
> Our perf tests should use the same code we recommend customers use. If customers need to handle 404 retry errors during delete/purge, we should either improve our SDK or samples, and then perf tests can align.

Agree that our samples need to be updated to show how to handle errors. In this case, we should check with @vcolin7 to understand what Key Vault library does when delete operations are retried and the secret is not found. If it swallows 404, then we don't need to update the sample or perf tests. If not, we need to update both.

vcolin7 commented:
> > Our perf tests should use the same code we recommend customers use. If customers need to handle 404 retry errors during delete/purge, we should either improve our SDK or samples, and then perf tests can align.
> 
> Agree that our samples need to be updated to show how to handle errors. In this case, we should check with @vcolin7 to understand what Key Vault library does when delete operations are retried and the secret is not found. If it swallows 404, then we don't need to update the sample or perf tests. If not, we need to update both.

After [having a look](https://github.com/Azure/azure-sdk-for-java/blob/897553254541f96236edf473fe789952180ab614/sdk/keyvault/azure-security-keyvault-secrets/src/main/java/com/azure/security/keyvault/secrets/SecretAsyncClient.java#L410) at how we handle `404` responses, it looks like the `beginDeleteSecret()` method will only throw a `ResourceNotFoundException` if a secret cannot be found at the beginning of the delete request ([the activation operation](https://github.com/Azure/azure-sdk-for-java/blob/897553254541f96236edf473fe789952180ab614/sdk/keyvault/azure-security-keyvault-secrets/src/main/java/com/azure/security/keyvault/secrets/SecretAsyncClient.java#L418) for endpoint `secrets/{secret-name}`). The SDK won't retry the to send the request in such cases.

All subsequent `404`s would come from a different endpoint (`deletedsecrets/{secret-name}`) and mean the delete operation [has not completed yet](https://github.com/Azure/azure-sdk-for-java/blob/897553254541f96236edf473fe789952180ab614/sdk/keyvault/azure-security-keyvault-secrets/src/main/java/com/azure/security/keyvault/secrets/SecretAsyncClient.java#L431). Other errors from the aforementioned endpoint will not throw but mark the operation [as completed instead](https://github.com/Azure/azure-sdk-for-java/blob/897553254541f96236edf473fe789952180ab614/sdk/keyvault/azure-security-keyvault-secrets/src/main/java/com/azure/security/keyvault/secrets/SecretAsyncClient.java#L439).
